### ISSUE TITLE
Multiple calls to `UI::LoadFont` can cause an assertion failure

### ISSUE THREAD

**Issue by Phlarx** [openplanet-nl/issues#39]
(Discovered in TMTurbo with version 1.7 of Ultimate Medals available here: https://github.com/openplanet-nl/plugin-signing/issues/150)

Repeated calls to `Resources::GetFont(fontFace, fontSize, -1, -1, true, true, true)` cause the following assertion failure:
```
[           Debug] [15:52:26]  Assertion failed: '"pTexture != 0"'
```

Steps to reproduce:
1. Load Ultimate Medals v1.7 in Turbo
2. Open settings for plugin to the "Display Settings" tab
3. Click anywhere on the Font Size slider
4. Wait for font to load
5. Repeat steps 3 and 4 until crash occurs (about 8 times for me)

Notes:
- The exact font size value of the final call does not seem to matter, as values that crash will load fine on next game boot
- I did not yet try changing font faces or within different games
- Does not affect Dashboard, since nvg is adjusting the size, not GetFont

**Comment by codecat** [comment]
Yes, your memory probably ran out. This is why you shouldn't be using a slider for stuff like this, but also because loading fonts takes a long time (and takes a lot of memory).

Right now, fonts aren't refcounted in scripts, so loading a font will always leak memory. This is on my list of things to fix.

**Comment by Phlarx** [comment]
I stated 'click the slider' instead of 'drag the slider' to eliminate the need to hit every intermediate value, suggesting it would still occur as quickly with any other numeric input.

This would be less urgent if there were a global scaling option, but I understand that it is hard. (And also I don't see an issue that currently asks for that...)

Anyway, is there a way to manually free a loaded font? I suppose that might equally be a recipe for disaster.

**Comment by codecat** [comment]
It's only hard because ImGui doesn't support on demand or vectorized font rendering yet.

I think I will make it refcount the fonts properly for this issue. Fonts can be shared between multiple plugins too, so s function to free a font wouldn't make sense.

**Comment by codecat** [comment]
I was just looking into this, and I'm not sure refcounting is the best idea either, since loading fonts can take quite a bit of time, which would make reloading plugins with custom fonts potentially really slow.. 😢

Still unsure what the correct solution would be. What would be ideal is [something like this](https://github.com/ocornut/imgui/pull/3471), but that might take a while to be implemented.

**Comment by codecat** [comment]
This issue remains difficult to solve. Even with proper refcounting, removing fonts from ImGui doesn't seem to be an easy task.

**Comment by codecat** [comment]
This 3 year old issue may get fixed soon! 👀 https://mastodon.gamedev.place/@ocornut/113687673612600791

**Comment by Fort-TM** [comment]
I assume the underlying issue on Imgui is the same? Because the issues are nothing alike otherwise

**Comment by codecat** [comment]
Upcoming dynamic fonts branch will fix this issue. I'll close this as I might push an early build of this to `edge` soon.
